src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -53,28 +54,36 @@ fn metadata_full(ws: &Workspace,
 
     Ok(ExportInfo {
         packages: packages,
-        resolve: Some(MetadataResolve(resolve, try!(ws.current()).package_id().clone())),
+        workspace_members: ws.members().map(|pkg| pkg.package_id().clone()).collect(),
+        resolve: Some(MetadataResolve{
+            resolve: resolve,
+            root: ws.current_opt().map(|pkg| pkg.package_id().clone()),
+        }),
         version: VERSION,
     })
 }
 
 #[derive(RustcEncodable)]
 pub struct ExportInfo {
     packages: Vec<Package>,
+    workspace_members: Vec<PackageId>,
     resolve: Option<MetadataResolve>,
     version: u32,
 }
 
 /// Newtype wrapper to provide a custom `Encodable` implementation.
 /// The one from lockfile does not fit because it uses a non-standard
 /// format for `PackageId`s
-struct MetadataResolve(Resolve, PackageId);
+struct MetadataResolve{
+    resolve: Resolve,
+    root: Option<PackageId>,
+}
 
 impl Encodable for MetadataResolve {
     fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
         #[derive(RustcEncodable)]
         struct EncodableResolve<'a> {
-            root: &'a PackageId,
+            root: Option<&'a PackageId>,
```

Comment:
I think this change is backwards compatible, because `root` would be null only in those cases, where previously `cargo metadata` did not work at all.